### PR TITLE
Suggest ref mut for pattern matching assignment

### LINKED ISSUE #118596: Suggestion to make a binding mutable could lead beginners astray
### Code

```Rust
fn main() {
    let y = Some(0);
    if let Some(x) = y {
        x = 2;
    }
}
```

### Current output

```Shell
error[E0384]: cannot assign twice to immutable variable `x`
 --> src/main.rs:4:9
  |
3 |     if let Some(x) = y {
  |                 -
  |                 |
  |                 first assignment to `x`
  |                 help: consider making this binding mutable: `mut x`
4 |         x = 2;
  |         ^^^^^ cannot assign twice to immutable variable
```

### Desired output

```Shell
error[E0384]: cannot assign twice to immutable variable `x`
 --> src/main.rs:4:9
  |
3 |     if let Some(x) = y {
  |                 -
  |                 |
  |                 first assignment to `x`
  |                 help: consider making this binding mutable: `mut x`
  |                 help: to modify the original value, take a borrow instead: `ref mut x`
4 |         x = 2;
  |         ^^^^^ cannot assign twice to immutable variable
```

### Rationale and extra context

In the original code, it's unclear whether the intention is to modify the contents of `y` or only the copy contained in `x`. The suggestion assumes the second case, which gives:
```rust
fn main() {
    let y = Some(0);
    if let Some(mut x) = y {
        x = 2;
    }
}
```
This doesn't modify `y` yet compiles fine, which I expect could be surprising to a beginner. This meaning of `mut` is far from obvious. With the updated error message, subsequent errors guide the user to actually modifying `y`.

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #118596

### CODE REVIEW

**Comment by petrochenkov** [comment]
r? compiler
I'm overloaded with reviews.

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -998,6 +998,17 @@ impl<'tcx> LocalDecl<'tcx> {
         )
     }
 
+    pub fn has_pattern_match(&self) -> bool {
```
Comment (wesleywiser) [inline compiler/rustc_middle/src/mir/mod.rs:1001]: Can you add a comment to this method?

File: compiler/rustc_middle/src/mir/mod.rs [hunk 2]
```diff
@@ -998,6 +998,17 @@ impl<'tcx> LocalDecl<'tcx> {
...
+        matches!(
+            self.local_info(),
+            LocalInfo::User(BindingForm::Var(VarBindingForm {
+                binding_mode: ty::BindingMode::BindByValue(Mutability::Not),
```
Comment (wesleywiser) [inline compiler/rustc_middle/src/mir/mod.rs:1005]: This seems more restrictive than the method name would imply. Can we check the mutability in the caller instead to make this more generic like the name?

Comment (chenyukang) [reply compiler/rustc_middle/src/mir/mod.rs]: yeah, it's not a general api.
I found the caller already has `decl.can_be_made_mutable()` to make it safe to add `mut`, so we don't need to check it.

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks @chenyukang!

@bors r+

**Comment by fmease** [comment]
Oh no, bors never caught the r+ :(

@bors r=wesleywiser

**Comment by fmease** [comment]
More than a month has passed, this might no longer build and might require a rebase. Therefore:

@bors rollup=iffy

**Comment by fmease** [comment]
Ah, I keep forgetting, I can actually retrigger CI, let's do that.

**Comment by fmease** [comment]
@chenyukang Could you rebase and fix CI? Thanks!
@rustbot author

**Comment by chenyukang** [comment]
> @chenyukang Could you rebase and fix CI? Thanks! @rustbot author

rebased and resolved tidy issues.

**Pull request merged**